src/gil.rs

```diff
@@ -451,7 +459,10 @@ pub unsafe fn register_incref(obj: NonNull<ffi::PyObject>) {
     if gil_is_acquired() {
         ffi::Py_INCREF(obj.as_ptr())
     } else {
+        #[cfg(feature = "reference-pool")]
         POOL.register_incref(obj);
+        #[cfg(not(feature = "reference-pool"))]
+        panic!("Cannot clone pointer into Python heap without the GIL being held.");
```

Comment:
In the spirit of #4098, should we add `#[track_caller]` here (and on the `Py::clone` impl, assuming that it's valid to add to trait methods)?

Reply:
It appears possible, c.f. https://rustc-dev-guide.rust-lang.org/backend/implicit-caller-location.html#traits and hence I will add it.

Reply:
I'll also add it to the `Drop` path because the backtrace reporting is just as relevant for the abort case.

Reply:
To be clear I'm not entirely sure aborting is necessary in Drop, as we already have precedent to leak e.g. unsendable objects. 

(I guess a further knob users may want a choice on is whether to leak or abort, but it's less clear that's a good fit for a feature.)

Reply:
FWIW, my concern with printing + leaking is that it's very difficult to catch in testing. If I want to ensure my app never takes these paths, it's much easier if I get a clear test failure.

Reply:
I agree with that but I think there is no right choice. Not even adding another knob seems just right, as the additional complexity seems over the top.

In the end, for me the consistency with how we already handle unsendable types was what tipped the odds towards leaking for me, because I think the problems are very much the same, even similarly niche, and if go for aborting here, we should do that for unsendable types as well.

Reply:
The case for testing is an interesting one, though.

What if we had an environment variable that upgraded leaking to abort (or vice versa) on debug builds for this case? I think the runtime overhead might be acceptable as part of a debugging setup.

I think for the unsendable case, aborting is not desirable at all because it's difficult to prevent the GC running on other threads.

Reply:
I would consider adding a `--cfg` which could be enabled via `RUSTFLAGS` etc. but personally, I find it too much complexity.

I mean, we all like performance, but for projects valuing correctness (e.g. cryptography), my recommendation would be to not keep doing what we have been doing until now. This is for projects which want to live on the razor's edge.

And in the long term, the problem will go away or rather move to a different project. A GILless CPython will carry the machinery to update reference counts from threads all over the place and we should be able to just unconditionally drop our reference pool for these builds.

Reply:
As discussed downthread, we have an additional `pyo3_leak_on_drop_without_reference_pool` conditional compilation flag now which switches from abort to leak on drop without the GIL.